File: 415 有效回文串/IsPalindrome.java
/**
 * Created by Silocean on 2016-04-04.
 */
public class IsPalindrome {
    public static void main(String[] args) {
        System.out.println(isPalindrome(".,"));
    }

    /**
     * @param s A string
     * @return Whether the string is a valid palindrome
     */
    public static boolean isPalindrome(String s) {
        // Write your code here
        s = s.toLowerCase();
        int i = 0;
        int j = s.length() - 1;
        while (i < j) {
            while (i < s.length() && !Character.isDigit(s.charAt(i)) && !Character.isLetter(s.charAt(i))) {
                i++;
            }
            while (j >= 0 && !Character.isDigit(s.charAt(j)) && !Character.isLetter(s.charAt(j))) {
                j--;
            }
            if (i < s.length() && j >= 0 && s.charAt(i) != s.charAt(j)) {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }

}

File: 415 有效回文串/IsPalindrome.py
def isPalindrome(s):
    s = s.lower()
    i = 0
    j = len(s) - 1
    while i < j:
        while i < len(s) and not s[i].isdigit() and not s[i].isalpha():
            i += 1
        while j >= 0 and not s[j].isdigit() and not s[j].isalpha():
            j -= 1
        if i < len(s) and j >= 0 and s[i] != s[j]:
            return False
        i += 1
        j -= 1
    return True

print isPalindrome('si 232 ,is')